back/node_watcher/src/nodeWatcher.ts

```diff
@@ -12,18 +12,48 @@ import { Cached, NomidotTask } from './tasks/types';
 
 const ARCHIVE_NODE_ENDPOINT =
   process.env.ARCHIVE_NODE_ENDPOINT || 'wss://kusama-rpc.polkadot.io/';
+const MAX_LAG = process.env.MAX_LAG || 0;
 
 const l = logger('node-watcher');
 
 function waitFinalized(
   api: ApiPromise,
   lastKnownBestFinalized: number
-): Promise<number> {
+): Promise<{ unsub: () => void; bestFinalizedBlock: number }> {
   return new Promise(resolve => {
     async function wait(): Promise<void> {
-      await api.derive.chain.bestNumberFinalized(best => {
+      const unsub = await api.derive.chain.bestNumberFinalized(best => {
         if (best.toNumber() > lastKnownBestFinalized) {
-          resolve(best.toNumber());
+          resolve({ unsub, bestFinalizedBlock: best.toNumber() });
```

Comment:
Right now we always unsub right after the query:
```ts
        const { unsub, bestFinalized } = await waitFinalized(api, blockIndex);
        unsub && unsub();
```

Instead of returning the unsub function and doing `unsub && unsub();`, could we just call the unsub() function inside waitFinalized?

```ts
          // Unsubscribe from api.derive.chain.bestNumberFinalized
          unsub(); 
          // Return the result we got
          resolve(best.toNumber());
```

and then simply:
```ts
const bestFinalized = await waitFinalized(api, blockIndex);
// no need to unsub here, it's done in the function
```

Basically, conceptually, waitFinalized creates the subscription, so it should be his job to clean it up too.

Reply:
I had try this, but it didn't work (unsub is undefined) because I was trying to call it right before the resolve, inside the subscription, actually that's what you suggest. This is a weird promise construction, I'll give it another try after lunch.

Reply:
Anything I tried so far didn't work, because of `undefined` unsub, or "`unsub` used before being assigned". So I'll leave it as is for now :(